controllers/dockermachine_controller.go

```diff
@@ -91,130 +123,70 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	// Always attempt to Patch the DockerMachine object and status after each reconciliation.
 	defer func() {
 		if err := patchHelper.Patch(ctx, dockerMachine); err != nil {
-			log.Error(err, "failed to patch DockerCluster")
-			if reterr == nil {
-				reterr = err
+			log.Error(err, "failed to patch DockerMachine")
+			if rerr == nil {
+				rerr = err
 			}
 		}
 	}()
 
+	// Handle deleted machines
+	if !dockerMachine.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(dockerMachine, machinesvc, lbsvc)
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(dockerMachine, machinesvc, lbsvc)
+}
+
+func (r *DockerMachineReconciler) reconcileNormal(dockerMachine *infrav1.DockerMachine, machinesvc *docker.Machine, lbsvc *docker.LoadBalancer) (ctrl.Result, error) {
 	// If the DockerMachine doesn't have finalizer, add it.
 	if !util.Contains(dockerMachine.Finalizers, infrav1.MachineFinalizer) {
 		dockerMachine.Finalizers = append(dockerMachine.Finalizers, infrav1.MachineFinalizer)
 	}
-	state := getState(machine, dockerMachine)
-	log = log.WithValues("state", state.String())
-	// TODO: consider setting the key pieces in this function, such as ProviderID and deleting finalizers
-	switch state {
```

Comment:
I'm ok with this going away...but can you tell me why you want to remove this? I find it much easier to read to make the current state explicit then act upon whatever state we're in

Reply:
+1

Reply:
There are three main reasons I think we should consider for this change:
1. afaik, many-core controllers in kubernetes do not rely on the human-readable state but on flags for controlling the lifecycle of objects (eg. Pod, Node)
2. also CAPA, which I consider the reference provider implementation today, does not rely on the human-readable state but on flags for controlling the lifecycle of objects (I understand there are differences among providers, but I'm still convinced that we all can benefit from more consistency whenever possible...)
3. IMO having a unique reconcileNormal cycle makes the code more suitable for managing external resources after creation, which is something we are not doing yet CAPD (e.g. recreate or restart containers).

Nevertheless, even if we are not using it for the controller logic, I will be happy to help in reusing this logic for implementing a human-readable state on the `DockerMachine.Status`

Reply:
I don't think the argument that "because that's how it's done elsewhere" is a great argument. It's easy to miss out on good ideas that way because you don't want to deviate from what already exists.

That being said, I think the implementation had some flaws. For example, Deleting should not have been part of the state, it was a special case and should have been handled before.

Thanks for writing this out. I might consider revisiting this idea again as I still think it's hard to understand what is actually happening in the reconciler code. In my mind we are using implicit state and it should be as explicit as possible.